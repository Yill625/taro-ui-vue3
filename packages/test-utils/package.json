{
  "name": "@taro-ui-vue3/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "dependencies": {
    "vue": "^3.0.10",
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}