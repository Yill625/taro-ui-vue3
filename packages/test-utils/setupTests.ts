import toHaveBeenWarnedInit from './been-warned'

toHaveBeenWarnedInit()