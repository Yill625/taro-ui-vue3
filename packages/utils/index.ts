import propsFactory from './propsFactory'

export * from './common'
export {
  propsFactory
}