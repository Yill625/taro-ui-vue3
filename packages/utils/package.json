{
  "name": "@taro-ui-vue3/utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "peerDependencies": {
    "@tarojs/taro": "^3.3.0",
    "vue": "^3.2.0"
  },
  "devDependencies": {
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}