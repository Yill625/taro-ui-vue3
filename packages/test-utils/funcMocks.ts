import * as utils from "@taro-ui-vue3/utils/common"
import { pxTransformMockFn, getEnvMockFn } from './@tarojs/taro'

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let pxTransform: jest.SpyInstance

function initMocks() {
  beforeEach(() => {
    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    pxTransform.mockRestore()
  })
}

export default initMocks