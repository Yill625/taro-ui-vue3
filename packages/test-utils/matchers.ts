import * as utils from "@taro-ui-vue3/utils/common"

jest.mock('@taro-ui-vue3/utils/common')

let warn: jest.SpyInstance
let uuid: jest.SpyInstance

const asserted: Set<string> = new Set()
const NOOP = () => { }

// Ensure console.warn is called with expected message
const toHaveBeenWarned: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed = warn.mock.calls.some(args => args[0].indexOf(received) > -1)
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned` +
        (msgs.length
          ? `.\n\nActual messages:\n\n - ${msgs}`
          : ` but no warning was recorded.`)
    }
  }
}

// Ensure previous console.warn is called with expected message
const toHaveBeenWarnedLast: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed =
    warn.mock.calls[warn.mock.calls.length - 1][0].indexOf(received) > -1
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned last.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned last.\n\nActual messages:\n\n - ${msgs}`
    }
  }
}

// Ensure console.warn is called n times with expected message
const toHaveBeenWarnedTimes: jest.CustomMatcher = (received: string, n: number) => {
  asserted.add(received)
  let found = 0
  warn.mock.calls.forEach(args => {
    if (args[0].indexOf(received) > -1) {
      found++
    }
  })

  if (found === n) {
    return {
      pass: true,
      message: () => `expected "${received}" to have been warned ${n} times.`
    }
  } else {
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned ${n} times but got ${found}.`
    }
  }
}

function initWarnMatchers() {
  beforeEach(() => {
    asserted.clear()
    warn = jest.spyOn(console, 'warn').mockImplementation(NOOP)
    uuid = jest.spyOn(utils, 'uuid').mockReturnValue('__tests__2021')

    // NOT working in concurrent mode
    // should avoid use in test.concurrent()
    expect.extend({
      toHaveBeenWarned,
      toHaveBeenWarnedLast,
      toHaveBeenWarnedTimes
    })
  })

  afterEach(() => {
    const assertedArray = Array.from(asserted)
    const nonAssertedWarnings = warn.mock.calls
      .map(args => args[0])
      .filter(received => {
        return !assertedArray.some(assertedMsg => {
          return received.indexOf(assertedMsg) > -1
        })
      })

    uuid.mockRestore()
    warn.mockRestore()

    if (nonAssertedWarnings.length) {
      throw new Error(
        `test case threw unexpected warnings:\n - ${nonAssertedWarnings.join(
          '\n - '
        )}`
      )
    }
  })
}

export default initWarnMatchers