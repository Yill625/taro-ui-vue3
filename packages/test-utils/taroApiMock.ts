import * as utils from "@taro-ui-vue3/utils/common"
import Taro from '@tarojs/taro'
import { createAnimation } from '@tarojs/taro-h5'

import {
  getEnvMockFn,
  pxTransformMockFn,
  chooseImageMockFn,
  querySelectorMockFn
} from './@tarojs/taro'

Taro.pxTransform = pxTransformMockFn
Taro.chooseImage = chooseImageMockFn
Taro.createSelectorQuery = querySelectorMockFn()
Taro.createAnimation = createAnimation

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let getenv: jest.SpyInstance
let pxTransform: jest.SpyInstance
let chooseImage: jest.SpyInstance

function initTaroMockApis() {
  beforeEach(() => {
    jest.mock('@tarojs/taro', () => {
      return Taro
    })

    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    getenv = jest
      .spyOn(Taro, 'getEnv')
      .mockImplementation(getEnvMockFn)

    chooseImage = jest
      .spyOn(Taro, 'chooseImage')
      .mockImplementation(chooseImageMockFn)

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    getenv.mockRestore()
    pxTransform.mockRestore()
    chooseImage.mockRestore()
  })
}

export default initTaroMockApis
