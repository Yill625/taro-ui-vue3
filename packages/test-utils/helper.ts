import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}