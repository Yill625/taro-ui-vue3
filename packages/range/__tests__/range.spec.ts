import { mountFactory, Slots } from '@taro-ui-vue3/test-utils/helper'
import AtRange from '../index'
import * as utils from '@taro-ui-vue3/utils/common'

const mountFn = (props?: object, slots?: Slots) => {
  return mountFactory(AtRange, undefined, props, slots)
}

describe('AtRange', () => {
  beforeEach(() => {
    jest.mock('@taro-ui-vue3/utils/common')
  })

  it('should render default AtRange', () => {
    const wrapper = mountFn()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render prop -- value', async () => {
    let delayQuerySelector = jest
      .spyOn(utils, 'delayQuerySelector')
      .mockReturnValue(new Promise((resolve) => {
        resolve([{ left: 20, width: 30 }])
      }))

    const wrapper = mountFn({
      value: [20, 50]
    })

    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    expect(
      wrapper.find('.at-range__track').attributes('style')
    ).toContain('left: 20%; width: 30%;')

    // wrapper.findAll('.at-range__slider').forEach((el, index) => {
    //   if (index === 0) {
    //     expect(
    //       el.attributes('style')
    //     ).toContain('left: 20%;')
    //   }

    //   if (index === 1) {
    //     expect(
    //       el.attributes('style')
    //     ).toContain('left: 50%;')
    //   }
    // })

    delayQuerySelector.mockRestore()
    delayQuerySelector = jest
      .spyOn(utils, 'delayQuerySelector')
      .mockReturnValue(new Promise((resolve) => {
        resolve([{ left: 30, width: 50 }])
      }))

    await wrapper.setProps({ value: [30, 80] })
    await wrapper.vm.$nextTick()
    expect(
      wrapper.find('.at-range__track').attributes('style')
    ).toContain('left: 30%; width: 50%;')

    // wrapper.findAll('.at-range__slider').forEach((el, index) => {
    //   if (index === 0) {
    //     expect(
    //       el.attributes('style')
    //     ).toContain('left: 30%;')
    //   }

    //   if (index === 1) {
    //     expect(
    //       el.attributes('style')
    //     ).toContain('left: 80%;')
    //   }
    // })

    delayQuerySelector.mockRestore()
  })

  it('should render prop -- sliderStyle', () => {
    const wrapper = mountFn({
      sliderStyle: 'color: red;'
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.at-range__slider').attributes('style')).toContain('color: red;')
  })

  it('should render prop -- railStyle', () => {
    const wrapper = mountFn({
      railStyle: 'color: red;'
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.at-range__rail').attributes('style')).toContain('color: red;')
  })

  it('should render prop -- trackStyle', () => {
    const wrapper = mountFn({
      trackStyle: 'color: red;'
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.at-range__track').attributes('style')).toContain('color: red;')
  })

  it('should render props -- min and max', async () => {
    const wrapper = mountFn({
      value: [20, 50],
      min: 10,
      max: 50
    })

    const aX = Math.round((20 - 10) / (50 - 10) * 100)
    const bX = Math.round((50 - 10) / (50 - 10) * 100)
    const style = `left: ${Math.min(aX, bX)}%; width: ${Math.abs(aX - bX)}%`

    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
    expect(
      wrapper.find('.at-range__track').attributes('style')
    ).toContain(style)
  })

  // PX will cause this test case fail, while 'px' won't
  it.skip('should render container and slider styles based on given blockSize', async () => {
    const wrapper = mountFn({ blockSize: 10 })

    await wrapper.vm.$nextTick()

    expect(
      wrapper.find('.at-range__container').attributes('style')
    ).toContain('height: 10PX;')

    wrapper.findAll('.at-range__slider').forEach(el => {
      expect(
        el.attributes('style')
      ).toContain('width: 10PX; height: 10PX; margin-left: -5PX;')
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('AtRange Behavior', () => {
  it('should trigger onChange by clicking on the track', async () => {
    const onChange = jest.fn()
    const wrapper = mountFn({
      onChange: onChange
    })
    await wrapper.find('.at-range__slider').trigger('touchend')
    await wrapper.find('.at-range').trigger('tap', {
      x: 30,
      y: 30
    })
    await wrapper.vm.$nextTick()

    expect(onChange).toBeCalled()
  })

  it('should not trigger onChange by clicking on the track when disabled', async () => {
    const onChange = jest.fn()
    const wrapper = mountFn({
      disabled: true,
      onChange: onChange
    })
    await wrapper.find('.at-range__slider').trigger('touchend')
    await wrapper.find('.at-range').trigger('tap', {
      x: 30,
      y: 30
    })
    await wrapper.vm.$nextTick()

    expect(onChange).not.toBeCalled()
  })

  it('should trigger onChange by touch moving the silder block', async () => {
    const onChange = jest.fn()
    const wrapper = mountFn({
      onChange: onChange
    })

    await wrapper.find('.at-range__slider').trigger('touchmove', {
      touches: [{
        clientX: 30,
        clientY: 30
      }]
    })
    await wrapper.vm.$nextTick()
    expect(onChange).toBeCalled()
  })

  it('should not trigger onChange by touch moving the silder block when disabled', async () => {
    const onChange = jest.fn()
    const wrapper = mountFn({
      disabled: true,
      onChange: onChange
    })

    await wrapper.find('.at-range__slider').trigger('touchmove', {
      touches: [{
        clientX: 30,
        clientY: 30
      }]
    })
    await wrapper.vm.$nextTick()
    expect(onChange).not.toBeCalled()
  })



  it('should trigger onAfterChange when touch ends', async () => {
    const onAfterChange = jest.fn()
    const wrapper = mountFn({
      onAfterChange: onAfterChange
    })

    await wrapper.find('.at-range__slider').trigger('touchend')

    expect(onAfterChange).toBeCalled()
  })

  it('should not trigger onAfterChange when disabled', async () => {
    const onAfterChange = jest.fn()
    const wrapper = mountFn({
      disabled: true,
      onAfterChange: onAfterChange
    })

    await wrapper.find('.at-range__slider').trigger('touchend')

    expect(onAfterChange).not.toBeCalled()
  })
})