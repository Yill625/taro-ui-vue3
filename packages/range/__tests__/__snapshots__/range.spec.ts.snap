// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtRange should render default AtRange 1`] = `
<view class="at-range">
  <view class="at-range__container">
    <view class="at-range__rail">
      <view class="at-range__track" style="left: 0%; width: 0%;"></view>
      <view class="at-range__slider" style="top: 0%;"></view>
      <view class="at-range__slider" style="top: 0%;"></view>
    </view>
  </view>
</view>
`;

exports[`AtRange should render prop -- railStyle 1`] = `
<view class="at-range">
  <view class="at-range__container">
    <view class="at-range__rail" style="color: red;">
      <view class="at-range__track" style="left: 0%; width: 0%;"></view>
      <view class="at-range__slider" style="top: 0%;"></view>
      <view class="at-range__slider" style="top: 0%;"></view>
    </view>
  </view>
</view>
`;

exports[`AtRange should render prop -- sliderStyle 1`] = `
<view class="at-range">
  <view class="at-range__container">
    <view class="at-range__rail">
      <view class="at-range__track" style="left: 0%; width: 0%;"></view>
      <view class="at-range__slider" style="color: red; top: 0%;"></view>
      <view class="at-range__slider" style="color: red; top: 0%;"></view>
    </view>
  </view>
</view>
`;

exports[`AtRange should render prop -- trackStyle 1`] = `
<view class="at-range">
  <view class="at-range__container">
    <view class="at-range__rail">
      <view class="at-range__track" style="color: red; left: 0%; width: 0%;"></view>
      <view class="at-range__slider" style="top: 0%;"></view>
      <view class="at-range__slider" style="top: 0%;"></view>
    </view>
  </view>
</view>
`;

exports[`AtRange should render prop -- value 1`] = `
<view class="at-range">
  <view class="at-range__container">
    <view class="at-range__rail">
      <view class="at-range__track" style="left: 20%; width: 30%;"></view>
      <view class="at-range__slider" style="top: 0%;"></view>
      <view class="at-range__slider" style="top: 0%;"></view>
    </view>
  </view>
</view>
`;

exports[`AtRange should render props -- min and max 1`] = `
<view class="at-range">
  <view class="at-range__container">
    <view class="at-range__rail">
      <view class="at-range__track" style="left: 25%; width: 75%;"></view>
      <view class="at-range__slider" style="top: 0%;"></view>
      <view class="at-range__slider" style="top: 0%;"></view>
    </view>
  </view>
</view>
`;
